File: Python/rotate-array.py
class Solution:
    # @param {integer[]} nums
    # @param {integer} k
    # @return {void} Do not return anything, modify nums in-place instead.
    def rotate(self, nums, k):
        def reverse(start, end):
            end -= 1
            while start<end:
                nums[start], nums[end] = nums[end], nums[start]
                start += 1
                end -= 1
                
        n = len(nums)
        k %= n
        
        reverse(0, n)
        reverse(0, k)
        reverse(k, n)

File: C++/rotate-array.cpp
class Solution {
private:
    void reverse(std::vector<int>& nums, int start, int end) {
        --end;
        while(start<end) {
            std::swap(nums[start], nums[end]);
            ++start;
            --end;
        }
    }
public:
    void rotate(std::vector<int>& nums, int k) {
        int n = nums.size();
        k %= n;
        reverse(nums, 0, n);
        reverse(nums, 0, k);
        reverse(nums, k, n);
    }
};

